File: G_Projects/G_Inventory_Management_System/ProductFactory/Product.java
package G_Projects.G_Inventory_Management_System.ProductFactory;

import G_Projects.G_Inventory_Management_System.CommonEnums.ProductEnum;

/*
 * Abstract Product Class - Key Features:
 * • Serves as the foundation for all product types in the inventory system
 * • Contains essential product attributes: SKU, name, price, quantity, and category
 * • Acts as a parent class that specific product implementations will extend
 * • Provides a common interface for managing different product varieties
 * • Ensures consistency across all product types through shared properties
 */

public abstract class Product{
    private String id;
    private String name;
    private double price;
    private int quantity;
    private int threshold;
    private ProductEnum category;

    public String getId() {
        return id;
    }
    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }
    public void setName(String name) {
        this.name = name;
    }

    public double getPrice() {
        return price;
    }
    public void setPrice(double price) {
        this.price = price;
    }

    public int getQuantity() {
        return quantity;
    }
    public void setQuantity(int quantity) {
        this.quantity = quantity;
    }

    public int getThreshold() {
        return threshold;
    }
    public void setThreshold(int threshold) {
        this.threshold = threshold;
    }

    public ProductEnum getCategory() {
        return category;
    }
    public void setCategory(ProductEnum category) {
        this.category = category;
    }
}

File: Python/A_Project/D_Inventory_Management_System/ProductFactory/product.py
class Product:
    def __init__(self):
        self._id = None
        self._name = None
        self._price = 0.0
        self._quantity = 0
        self._threshold = 0
        self._category = None

    def get_id(self):
        return self._id

    def set_id(self, product_id: str):
        self._id = product_id

    def get_name(self):
        return self._name

    def set_name(self, name: str):
        self._name = name

    def get_price(self):
        return self._price

    def set_price(self, price: float):
        self._price = price

    def get_quantity(self):
        return self._quantity

    def set_quantity(self, quantity: int):
        self._quantity = quantity

    def get_threshold(self):
        return self._threshold

    def set_threshold(self, threshold: int):
        self._threshold = threshold

    def get_category(self):
        return self._category

    def set_category(self, category):
        self._category = category
